Complex navigator first version

jmedinaebi opened the pull request:
All the changes for the 1st releasable version of the Complex Navigator.

susiehgt committed:
Matrix created

The matrix has been created, the component added in it but they are not working with the complexes

susiehgt committed:
Delete complex-results.component.ts

susiehgt committed:
Delete complex-list.component.html

susiehgt committed:
Delete complex-list.component.css

susiehgt committed:
Matrix created

The matrix has been created, the component added in it but they are not working with the complexes

susiehgt committed:
Matrix created

The matrix has been created, the component added in it but they are not working with the complexes

susiehgt committed:
Update complex-list.component.html

Corrections of the code

susiehgt committed:
Adding of some checkings

susiehgt committed:
Adding of Hack method

susiehgt committed:
Update on the progression of the complex portal (adding of the data in the matrix)

susiehgt committed:
Update on the progression of the complex portal (adding of the data in the matrix)

susiehgt committed:
Update

The last version was not updated correctly

jmedinaebi committed:
Add links to components

jmedinaebi committed:
Only display external link if there is one

susiehgt committed:
Adding of the subcomplexes components, not finished (the table is not yet adapted and the stoichiometry has to be added)

susiehgt committed:
Collapsible menu and stoichiometry of subcomplexes added but viewing not optimized , modification of the headers and how complexes' names are displayed

susiehgt committed:
-displaying of the collapsible menus fixed.
-stoichiometry of subcomplexes has to be fixed (display 1 when element here but not the real stoichiometry)

susiehgt committed:
-stoichiometry of the subcomplexes fixed
-displaying of the name and function of the interactors
-minor fixings on the table display
NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
-fixing of the view button
NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
- fixing the stoichiometry: now the main complex column also displays the stoichiometry of the interactors contained in the subcomplexes
NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
- fixing the display of the subcomplexes, not it only shows the one selected
NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
Creation of complex-navigator directory
NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
Creation of complex-navigator directory
NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
Creation of complex-navigator directory
NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
Complex-navigator functional (except for see notice)
Display fixed

NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
Fixing of some elements of the display (icons not showing properly, size of the cells being different)

NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
Modification of the display of the stoichiometry
Fixing of the magnifying glass button
Changing of the display of the interactors' names
Correction of some coding elements

NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

jmedinaebi committed:
Remove extra calls to fetch components, now returned from BE

susiehgt committed:
Adding of the connecting lines (not finished)
Changes on the display

NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
Merge branch 'complex-navigator' into complex-navigator-backend-changes

susiehgt committed:
Merge pull request #208 from Complex-Portal/complex-navigator-backend-changes

Remove extra calls to fetch components, now returned from BE

jmedinaebi committed:
Rename components as interactors

jmedinaebi committed:
Make table taller

jmedinaebi committed:
Refactor table interactor column to hide duplicated components

jmedinaebi committed:
Revert "Refactor table interactor column to hide duplicated components"

This reverts commit f531c545455d96890aab46507a4efaeaaf531df9.

jmedinaebi committed:
Rename components as interactors

jmedinaebi committed:
Simplify use of ngIfs with ng-container

jmedinaebi committed:
Create EnrichedInteractor class to store metadata about interactors displayed

jmedinaebi committed:
Refactor methods to get stochiometry

jmedinaebi committed:
Hide interactors already displayed in expanded section to avoid duplicated rows

jmedinaebi committed:
Fetch interactors in subcomplex from backend if needed

susiehgt committed:
Merge pull request #209 from Complex-Portal/complex-navigator-backend-changes

Complex navigator backend changes

susiehgt committed:
Adding of the connection lines between the circles

(missing participants not close to each other)

NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

susiehgt committed:
Adding of the connection lines between the circles

NOTICE: the subcomplex expanding doesn't work on the research results where the subcomplexes are not also listed as a complex

jmedinaebi committed:
Use ngClass to draw the lines for the interactors

jmedinaebi committed:
Remove unused parameter

susiehgt committed:
Merge pull request #210 from Complex-Portal/complex-navigator-use-ng-class

Complex navigator use ng class

susiehgt committed:
Aesthetic corrections, alignement of the vertical lines

jmedinaebi committed:
Pre-calculate start and end indices

jmedinaebi committed:
Refactor to make logic more readable

jmedinaebi committed:
Refactor for readability

jmedinaebi committed:
Remove unused import

susiehgt committed:
Aesthetic corrections, alignement of the vertical lines

jmedinaebi committed:
Merge branch 'complex-navigator' into complex-navigator-use-ng-class

jmedinaebi committed:
Small change in CSS

susiehgt committed:
Adding of some icons, modification of the display of the complexes

susiehgt committed:
Merge pull request #211 from Complex-Portal/complex-navigator-use-ng-class

Pre-calculate start and end indices

jmedinaebi committed:
Fix wrong FTP url

jmedinaebi committed:
New npm command to publish github pages with test env

susiehgt committed:
Changing of the display of the complexes

susiehgt committed:
fix of the length of the complex name

jmedinaebi committed:
Fix issues with lines connecting interactors

susiehgt committed:
Merge pull request #212 from Complex-Portal/complex-navigator-fix-lines

Fix issues with lines connecting interactors

susiehgt committed:
Display of overflowing table header changed, still not quite optimised depending on screen size.
Adding of icons on the filter list

susiehgt committed:
Merge remote-tracking branch 'origin/complex-navigator' into complex-navigator

susiehgt committed:
Classification of the complexes depending on their similarity almost okay (need to check).
Classification of the interactors on the way

susiehgt committed:
Classification of interactors ok but lines not ok anymore

adding of line between complexes and interactors column
some fixings of the display (height and width of the matrix) -> Need to check for the stickiness of the header

susiehgt committed:
fixing of some display, header sticky but table expands to the right

susiehgt committed:
Sorting of the complexes depending on how much similar components they have with the biggest;
Changing of table display

susiehgt committed:
Short lines algorithm in progress

susiehgt committed:
Short length algorithm working on little search but crashing on big search

jmedinaebi committed:
Rename components as interactors

susiehgt committed:
Adding of a new object for binary array

susiehgt committed:
Merge pull request #213 from Complex-Portal/navigator-rename-components-as-interactors

Rename components as interactors

susiehgt committed:
Delete src/app/complex/complex-results/complex-navigator/table-structure/table-header/table-header-style2.component.css

useless

susiehgt committed:
BinaryList updated

susiehgt committed:
Sorting of the complexes by similarties OK; sorting of the components still have to include stoichiometry of subcomplexes

susiehgt committed:
Merge remote-tracking branch 'origin/complex-navigator' into complex-navigator

# Conflicts:
#	src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.ts

susiehgt committed:
Classifying interactors by appearance ok

susiehgt committed:
Adding of the ranges for interactors types and organism;
Button for choosing the types of listing of the interactors not functional, but functions work separatly

susiehgt committed:
Fixing of the sticky header and the clickability of the complexes header

susiehgt committed:
Beginning of adding lines depending on interactors listing

susiehgt committed:
Interactors classification button + sidebars OK

susiehgt committed:
Fixing of the size of the
table and scrolling issues caused by it;
Fixing of typos;
Interactors Occurence sorting button to check (scrolling refreshing makes the whole table lagging by playing the sorting function);
Sorting buttons to check (have to click multiple time to work)

susiehgt committed:
Ranges of interactor type V2, not working when expanded

susiehgt committed:
Ranges of interactor type V2, working but some bugs on the expanded complex

susiehgt committed:
Ranges of interactor type V2, working but some bugs on the first expanded complex

susiehgt committed:
Modification of the classification divs

susiehgt committed:
Fixing of some classification bugs

susiehgt committed:
Fixing of interactors displaying

susiehgt committed:
code reorganisation and optimisation, fixing of display elements

susiehgt committed:
Creation of complex portal utils, changing of icons function.
Start to implement the adjusting of the interactors header to the size of the cells when sorting is on §§§§

susiehgt committed:
Creation of complex portal utils, changing of icons function.
Start to implement the adjusting of the interactors header to the size of the cells when sorting is on §§§§

susiehgt committed:
Removing of the tslint comments

susiehgt committed:
Removing of useless code, fixing of some duplications

susiehgt committed:
Calling from backend changed, removing of useless imports and ngOnInit

susiehgt committed:
Fixing of displaying bugs while using filters and/or changing display type,
adding of the view in the IRL,
adding of the icons for the species,
adding of pagination tools (first/last page)

susiehgt committed:
Testing types of display

susiehgt committed:
Fixing of display for new options, now working correctly

susiehgt committed:
Display working correctly;
Sorting complexes function modified, now working correctly;
Cell height back to previous to be in tune with the scrolling function

susiehgt committed:
Fixing of double stoichiometry number appearing;
Using of external link class;
Fixing of minor displaying issues

susiehgt committed:
If one complex is searched, the result is the complex portal page

jmedinaebi commented on src/app/complex/complex-results/complex-filter/complex-filter.component.ts:
```diff
@@ -1,6 +1,8 @@
 import {Component, EventEmitter, Input, OnInit, Output} from '@angular/core';
 import {Facets} from '../../shared/model/complex-results/facets.model';
 import {AnalyticsService} from '../../../shared/google-analytics/service/analytics.service';
+import {Interactor} from '../../shared/model/complex-results/interactor.model';
```
Are these import needed?

jmedinaebi commented on src/app/complex/complex-results/complex-navigator-sorting/complex-navigator-sorting.component.html:
```diff
@@ -0,0 +1,49 @@
+<div class="filters">
+
+  <div class="sortingInteractors">
```
We should either split this into 2 components, or rename the component.

It's called complex-navigator-sorting at the moment, and it contains 2 things, the menu for sorting, and then the menu for the view; so the name is slightly wrong now.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-subcomponent-interactor/table-subcomponent-interactor.component.ts:
```diff
@@ -0,0 +1,163 @@
+  subComplexInteractorStoichiometryValue: string;
+
+
+  // findInteractorInComplex = findInteractorInComplex;
```
Commented out lines, please delete.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-subcomponent-interactor/table-subcomponent-interactor.component.html:
```diff
@@ -0,0 +1,12 @@
+<div [ngClass]="displayTopLineClass"></div>
+<cp-table-interactor-stoichiometry
+  *ngIf="interactorInComplex"
+  [stoichiometryText]="interactorStoichiometryText"
```
Not completely necessary, but the same thing we did in this component, having the values in variables set in ngOnChangesn and not calling methods in ngIf; we could do in other components too.

jmedinaebi commented on src/app/complex/complex-results/complex-results.component.ts:
```diff
@@ -194,4 +221,40 @@ export class ComplexResultsComponent implements OnInit, AfterViewInit {
+    this.reloadPage();
+  }
+
+  setFirstDisplayType(): void {
```
`window.location.href` is not the proper way to redirect to the complex details page. Also, you're using localhost, which worked for testing locally, but it won't remote.

Try replacing it with
```
  setFirstDisplayType(): void {
    if (this._complexSearch.elements.length === 1) {
      const complexId = this._complexSearch.elements[0].complexAC;
      if (!!complexId) {
        // For some reason this is needed so the navigate call works
        this.router.routeReuseStrategy.shouldReuseRoute = function () {
          return false;
        };
        this.router.navigate(['/complex', complexId]);
      }
    }

    if (!this.DisplayType) {
      // Currently the list view is the default, as we are just launching the navigator view
      // Later on we can change the default view to be the list or navigator view based on number of results
      this.setListView();
    }
  }
```

I also changed the logic of the display type to have list be default for now, so when we first launch it we keep list as the default view. Maybe for the UX testing, we can have the navigator view as default, and then switch it back.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-name/table-interactor-name.component.html:
```diff
@@ -0,0 +1,36 @@
+
+    <a [routerLink]="['/complex/search']"
+       [queryParams]="{query: interactorId, page: 1}"
+       [fragment]="'view_complex_navigator'"
```
Instead of `[fragment]="'view_complex_navigator'"`, we can use `[preserveFragment]="true"` to keep the same fragment we currently have in the URL. That way, if we rename it, then it does not matter.

jmedinaebi commented:
One more thing I noticed, the sorting menu goes behind the complex navigator table

![Screenshot 2024-05-10 at 09 47 15](https://github.com/Complex-Portal/complex-portal-view/assets/114929382/d23f1a7f-f123-4f46-9196-aa32900ad296)

It does not happen with the view menu

![Screenshot 2024-05-10 at 09 47 19](https://github.com/Complex-Portal/complex-portal-view/assets/114929382/658a5f38-77c8-409a-a42a-becbe9b388b4)

susiehgt committed:
Change of file names, deletion of useless parts

susiehgt committed:
Fixing on buttons display

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.css:
```diff
@@ -118,3 +118,7 @@ a, a:visited {
 }
+
+.container {
+  font-size: 14px;
```
What's this change for?

susiehgt replied:
It is the size of the expandable button which was bigger than the rest

jmedinaebi commented on src/app/complex/complex-results/complex-results.module.ts:
```diff
@@ -27,7 +27,7 @@ import {
   TableSubcomponentInteractorComponent
 } from './complex-navigator/table-structure/table-interactor-column/table-subcomponent-interactor/table-subcomponent-interactor.component';
-import {ComplexNavigatorSortingComponent} from './complex-navigator-sorting/complex-navigator-sorting.component';
+import {ComplexNavigatorSortingComponent} from './complex-navigator-buttons/complex-navigator-sorting.component';
```
You rename the component selector `cp-complex-navigator-buttons` and rename the folder containing the component `complex-navigator-buttons`; but the actual component remains with the old name.

Files inside ``complex-navigator-buttons`:
- complex-navigator-sorting.component.ts
- complex-navigator-sorting.component.spec.ts
- complex-navigator-sorting.component.css
- complex-navigator-sorting.component.html

And the component is still called `ComplexNavigatorSortingComponent`.

Nothing important but we should rename all of this for consistency.

susiehgt committed:
Adding of icons in the complexes,
Fixing of display of sorting interactors column (it was not sticking correctly)

susiehgt committed:
Fixing of display of sorting subinteractors column (it was not sticking correctly)

susiehgt committed:
Fixing icon display which were not working

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-header/table-header.component.html:
```diff
@@ -0,0 +1,66 @@
+      </th>
+      <div class="spaceHolder"></div>
+    </tr>
+    <tr [style.--cols]="complexes.length" [style.--intHeader]="isInteractorSortingSet() ? '156px':''">
```
Should we have `*ngIf="organismIconDisplay"` here? Not sure about it, but if we add it, we then also need to add it in line 18.

susiehgt committed:
Interactor organism now fetched from backend;
Formating organism string function moved to utils;
adding of organism icons in the filters;

susiehgt committed:
Moving the filters to the left;
Making the complex navigator fit the page;
Centering of the buttons

susiehgt committed:
removing of useless functions

susiehgt committed:
fixing of header display

susiehgt committed:
Fixing interactors display,
fixing filters display

susiehgt committed:
Adding user review notification

susiehgt committed:
Adding of "new view" pop up

jmedinaebi commented on src/app/shared/notification/service/notification.service.ts:
```diff
@@ -45,4 +51,12 @@ export class NotificationService {
   }
+
+  public rateComplexNavigator(): void {
+    this.addRatingNotification('<div><input class="input-small" value="textbox"/><a href="' + environment.complex_portal_support_url + '">Tell us about your experience with the Complex Navigator!</a></div>');
```
Add `target="_blank"` to the `<a>` element `<a target="_blank" href=...`, so it opens a new tab.

jmedinaebi commented on src/app/shared/notification/service/notification.service.ts:
```diff
@@ -45,4 +51,12 @@ export class NotificationService {
+    this.addRatingNotification('<div><input class="input-small" value="textbox"/><a href="' + environment.complex_portal_support_url + '">Tell us about your experience with the Complex Navigator!</a></div>');
+  }
+
+  public complexNavigatorAnnouncement() {
```
At the moment the pop up displays at the very top, and it's easy to miss. We can configure it to display lower so it displays on the white part of the screen.

1. In `styles.css` add the following class:
```
.toast-right-under-header {
  right: 10px;
  top: 225px;
}
```

2. Update this method to use that position class:
```
  public complexNavigatorAnnouncement() {
    this.addAnnouncementNotification(
      'Try out our new display: the Complex Navigator!',
      {
        positionClass: 'toast-right-under-header'
      });
  }
```

Give it a try, and you can also use it for the `rateComplexNavigator` if you want.

jmedinaebi replied:
There are more options, like `disableTimeOut` and `closeButton`. So we can disable the timeout and force users to have to close the pop up.

susiehgt committed:
Adding of the rating component with google Analytics (not functional)

jmedinaebi commented on src/app/shared/notification/service/notification.service.ts:
```diff
@@ -53,10 +55,26 @@ export class NotificationService {
+    this.addAnnouncementNotification(
+      'Try out our new display: the Complex Navigator!',
+      {
+        positionClass: 'toast-right-under-header'
```
So it doesn't close automatically and people keep seeing it, add the following options:
```
        positionClass: 'toast-right-under-header',
        disableTimeOut: true,
        closeButton: true
```

jmedinaebi commented on src/app/shared/notification/service/notification.service.ts:
```diff
@@ -53,10 +55,26 @@ export class NotificationService {
+    <input type="submit" name="submitButton" value="Submit Rating" class="button" (click)="submitRating()">
+  </div>
+    `;
+    this.addRatingNotification(ratingHtml);
```
I've tried but I haven't managed to make the buttons inside the toast clickable. I even used `toastComponent` and created a new component but no luck. I did a quick search and it may be that this type of toast messages don't work with clickable buttons in it.

susiehgt committed:
Adding of a popup for rating and its own class;
removing of the rating notification which was not functional

susiehgt committed:
Fixing the notification display

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/complex-navigator-rating/complex-navigator-rating.component.ts:
```diff
@@ -0,0 +1,33 @@
+  }
+
+  submitRating() {
+    if (this.selectedRating !== null) {
```
Replace with
```
if (!!this.selectedRating) {
```

Initially `selectedRating` is undefined, not null (typescript differentiate between these two values), so because selectedRating is not null, it goes in this if and tries to do selectedRating.toString(), which results in 
```
ERROR TypeError: Cannot read properties of undefined (reading 'toString')
    at ComplexNavigatorRatingComponent.submitRating (complex-navigator-rating.component.ts:23:70)
```

`!!this.selectedRating` is equivalent to write `this.selectedRating != null and this.selectedRating != undefined`.

The only thing that can be tricky with !! is that it is more than just null and undefined, 0 may be also one of the values considered. But, because we don't have a 0 rating value, then it's ok.

jmedinaebi commented on src/app/shared/notification/service/notification.service.ts:
```diff
@@ -25,6 +25,14 @@ export class NotificationService {
   }
 
+  private addComplexNavigatorAnnouncementNotification(announcement: string): void {
+    this.toastrService.toastrConfig.disableTimeOut = true;
```
I think when we do `this.toastrService.toastrConfig.disableTimeOut = true;`, we're changing the config for all toasts, not just for this announcement, so if any other toast gets displayed, it will also not have the timeout.

The way to do it only for this is
```
    return this.toastrService.info(announcement, null, {
      disableTimeOut: true,
      closeButton: true,
      progressBar: false,
      positionClass: 'toast-right-under-header'
    });
```
2nd argument is null as it is the title, which we do not set.

jmedinaebi commented on src/app/shared/notification/service/notification.service.ts:
```diff
@@ -45,4 +53,14 @@ export class NotificationService {
+      'the Complex Navigator!');
+  }
+
+  public closeAnnouncement(): void {
```
As discussed, updated `addComplexNavigatorAnnouncementNotification` to return the result the toast.info call 
```
return this.toastrService.info(...);
```

Then, update `complexNavigatorAnnouncement` to also return
```
return this.addComplexNavigatorAnnouncementNotification(...)
```

Then, in complex-results.component.ts, store the returned value
```
this.toast = this.notificationService.complexNavigatorAnnouncement();
```

Update `closeAnnouncement` to take the toast id
```
  public closeAnnouncement(toastId: number): void {
    this.toastrService.clear(toastId);
  }
```

And finally, update the call to it
```
setComplexNavigatorView() {
    this.DisplayType = this.COMPLEX_NAVIGATOR_VIEW;
    this.reloadPage();
    if (!!this.toast) {
      this.notificationService.closeAnnouncement(this.toast.toastId);
      this.toast = null;
    }
  }
```

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/complex-navigator-rating/complex-navigator-rating.component.ts:
```diff
@@ -0,0 +1,33 @@
+  submitRating() {
+    if (this.selectedRating !== null) {
+      this.analyticsService.rateComplexNavigator(this.selectedRating.toString());
+      console.log(`Submitting rating: ${this.selectedRating}`);
```
We need to close this popup when the rating is submitted.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator-buttons/complex-navigator-buttons.component.ts:
```diff
@@ -0,0 +1,74 @@
+
+  DisplayingOrganism() {
+    this.organismIconDisplay = !this.organismIconDisplay;
+    this.changesEmitter();
```
If `organismIconDisplay` is set to true, we need to set detailed display to true too. It may already be true, but it doesn't hurt setting it anyway.

```
    this.organismIconDisplay = !this.organismIconDisplay;
    if (this.organismIconDisplay) {
        this.IDDisplay = true;
    }
    this.changesEmitter();
```

Same change needed in `DisplayingType` and `DisplayingID` with the relevant variables.

This is to fix the issue of, being in compact display and clicking on one of the detailed checkboxes.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.html:
```diff
@@ -0,0 +1,97 @@
+        <ng-container *ngIf="!!interactor.subComponents">
+          <tr *ngFor="let el of interactor.subComponents; let j=index"
+              class="expandedRows"
+              [ngClass]="{'firstExpandedRow': j === 0, 'lastExpandedRow': j === interactor.subComponents.length - 1}">
```
The issue with the cells not having the right width with a low number of complexes is in this line. I'm still looking into it, but this is definitely the line.

jmedinaebi replied:
I think I got it.

In `table-interactor-column.component.css`, for class `interactorsColumn`, get rid of `width: 130px;` (line 16); and for class `intStoich`, get rid of `width: 70px;` (line 32).

susiehgt replied:
I think I've put it this way because of the interactors header which was not sizing the right way but I will check that tomorrow because it was working well on the previous versions

jmedinaebi replied:
I managed to fix this issue.

1. You may have some CSS changes locally, revert them as what you have here is not far from working.

2. Update classes in table-interactor-column.component.css
    ```
    .interactorsColumn {
      position: sticky;
      left: 0;
      z-index: 3;
      max-height: 45px;
      min-height: 45px;
      min-width: 130px;
      max-width: 130px;
      padding: 5px;
      text-align: end;
      color: #0e6f76;
      font-weight: bold;
      background: inherit;
    }
    
    .interactorsColumnFixedWidth {
      width: 130px;
    }
    
    .intStoich {
      height: 45px;
      max-width: 70px;
      min-width: 70px;
      font-size: medium;
      text-align: end;
      z-index: 1;
    }

    .intStoichFixedWidth {
      width: 70px;
    }
    ```

3. Update html to use new CSS classes
    3.1. lines 15-16
    ```
    <td class="interactorsColumn" [ngClass]="complexes.length<=8 ? 'horizontalHeader' : 'interactorsColumnFixedWidth'"
            [style]="ranges.length!=0 ? 'left:3ch':''">
    ```
    3.2. line 39
    ````
    <td class="intStoich" [ngClass]="{'intStoichFixedWidth': complexes.length > 8}">
    ````
    3.3. line 82
    ```
    <td class="intStoich" [ngClass]="{'intStoichFixedWidth': complexes.length > 8}">
    ```
    3.4. also line 91, not classes but there is a bug there
    ```
    <div class="spaceHolder" *ngIf="complexes.length>8"></div>
    ```

jmedinaebi replied:
I tested it with the searches
- nuclear pore
- KMN: 9 results
- CPX-4441: 3 results

jmedinaebi commented:
I've tested how to handle pagination for list view and complex view separately and it's easier than I thought. We don't need to move the paginator to the list and complex navigator components. We just need to rely on setters and getters to set or get the right value.

1. Delete the current variables to store pagination values
    ```
    private _currentPageIndex: number;
    private _lastPageIndex: number;
    private _pageSize = 15;
    ```

2. Add new variables, having separate ones for list and navigator view
    ```
      private _listPageSize = 15; // This is where we set the size of the pages for list view
      private _navigatorPageSize = 30; // This is where we set the size of the pages for navigator view
      private _listCurrentPage: number;
      private _navigatorCurrentPage: number;
      private _listLastPageIndex;
      private _navigatorLastPageIndex;
    ```

3. Update setters and getters for `currentPageIndex`, `lastPageIndex` and `pageSize`. They would set or get the right variable based on the display type currently set.
    ```
      get currentPageIndex(): number {
        if (this.DisplayType === this.COMPLEX_NAVIGATOR_VIEW) {
          return this._navigatorCurrentPage;
        } else {
          return this._listCurrentPage;
        }
      }
    
      set currentPageIndex(value: number) {
        if (this.DisplayType === this.COMPLEX_NAVIGATOR_VIEW) {
          this._navigatorCurrentPage = value;
        } else {
          this._listCurrentPage = value;
        }
      }
    
      get lastPageIndex(): number {
        if (this.DisplayType === this.COMPLEX_NAVIGATOR_VIEW) {
          return this._navigatorLastPageIndex;
        } else {
          return this._listLastPageIndex;
        }
      }
    
      set lastPageIndex(value: number) {
        if (this.DisplayType === this.COMPLEX_NAVIGATOR_VIEW) {
          this._navigatorLastPageIndex = value;
        } else {
          this._listLastPageIndex = value;
        }
      }
    
      get pageSize(): number {
        if (this.DisplayType === this.COMPLEX_NAVIGATOR_VIEW) {
          return this._navigatorPageSize;
        } else {
          return this._listPageSize;
        }
      }
    ```

4. Update any part of the code using the variables we deleted in step 1 directly instead of the setters or getters. No changes are needed in the HTML, but some are needed in the ts file.

    4.1. In `ngOnInit` method, 5th line inside the `this.route.queryParams.subscribe` body.
    ```
    this.currentPageIndex = queryParams['page'] ? Number(queryParams['page']) : 1;
    ```
    4.2. In `reloadPage` method, 3rd line of the method
    ```
    queryParams['page'] = this.currentPageIndex;
    ```
    Other parts of the code, like setting the last page, already use the setters and getters, so no more changes are needed.

jmedinaebi commented:
Regarding my latest comment about the pagination changes, there is one issue. When the page is reloaded, the body in `this.route.queryParams.subscribe` runs before the body in `this.route.fragment.subscribe`. This means that we try to do the search, and set pagination variables, before we set the display.

We can fix it but having one subscribe inside the other. Instead of
```
this.route.queryParams.subscribe(queryParams => {
    ...
});

this.route.fragment.subscribe(fragment => {
    ...
});
```
As we currently have it, we can do
```
this.route.fragment.subscribe(fragment => {
    ...
    this.route.queryParams.subscribe(queryParams => {
        ...
    });
});
```

Then, we make sure the display type is set before we parse the query params.

susiehgt committed:
Fixing the display buttons,
Fixing the header size bug,
changing of the view depending on the number of search results

jmedinaebi commented on src/app/complex/complex-results/complex-results.component.ts:
```diff
@@ -194,4 +242,47 @@ export class ComplexResultsComponent implements OnInit, AfterViewInit {
+        };
+        this.router.navigate(['/complex', complexId]);
+      }
+    } else if (this._complexSearch.elements.length < this._navigatorPageSize) {
```
A few things here.

1. `this._complexSearch.elements` are the elements in the page, so they will always be 15 or 20. We should do 
```
if (this._complexSearch.totalNumberOfResults <= this._navigatorPageSize) {
```

2. As you can see in the code above, we should use `<=` instead of `<`. If there are 20 results or less, we want the navigator view.

3. More importantly, the way this is set up, we always check the length of the elements after a search. That means that if someone goes to `http://localhost:4200/complex/search?query=KMN&page=1#view_list`, which has the fragment `#view_list`, because there are 9 results, they get redirected to the navigator view. We should only do this if the fragment is not set, which happens for urls like `http://localhost:4200/complex/search?query=KMN&page=1` or when people initially search. To fix this we need to move this if/else inside the if below:
```
    if (!this.DisplayType) {
      if (this._complexSearch.totalNumberOfResults <= this._navigatorPageSize) {
        this.setComplexNavigatorView();
      } else {
        this.setListView();
      }
    }
```

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-name/table-interactor-name.component.html:
```diff
@@ -0,0 +1,36 @@
+
+    <a [routerLink]="['/complex/search']"
+       [queryParams]="{query: interactorId, page: 1}"
+       [preserveFragment]="true"
```
Because we now check the number of results to decide between list or navigator view, it makes sense to get rid of this line. So, instead preserving the fragment, when people click on search of one of the interactors, it opens a new tab, and that tab is list or navigator view depending on the number of results.

jmedinaebi commented on src/app/complex/complex-results/complex-results.component.ts:
```diff
@@ -5,39 +5,56 @@ import {ComplexPortalService} from '../shared/service/complex-portal.service';
+        this.DisplayType = this.LIST_VIEW;
+      }
+    });
     this.route
```
Here I didn't mean to move `this.route.queryParams.subscribe` after `this.route.fragment.subscribe`. I meant moving it inside. The potential issue with the 2 subscribers separate is that they are synchronous, so we cannot make sure of the order they run, so we may be checking the query params before the fragment and we shouldn't as there are variables, like page index, that depend on the fragment. So, to avoid that we can do

```
    this.route.fragment.subscribe(fragment => {
      if (fragment === this.COMPLEX_NAVIGATOR_VIEW) {
        this.DisplayType = this.COMPLEX_NAVIGATOR_VIEW;
      } else if (fragment === this.LIST_VIEW) {
        this.DisplayType = this.LIST_VIEW;
      }
      this.route
        .queryParams
        .subscribe(queryParams => {
            ...
        });
    });
```

So, after processing the fragment, we process the query params.

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/complex-navigator-rating/complex-navigator-rating.component.html:
```diff
@@ -0,0 +1,15 @@
+<div class="ratingComponent" *ngIf="visible===true">
+  <div class="text">
+    <p>Tell us about your experience with the Complex Navigator!</p>
+    <button class="close" id="closePopup" (click)="ratingVisibility()" style="color: white">&times;</button>
```
The 'x' to close the pop up is not positioned right, it should be in the top right corner, and it's next to the text.

![Screenshot 2024-05-23 at 14 01 45](https://github.com/Complex-Portal/complex-portal-view/assets/114929382/697b4ca3-e3cf-4e03-a7d2-0743e03571e3)

susiehgt committed:
Fixing of rating display,
removing of useless function,
modification of the url function

susiehgt committed:
Removing of commented parts,
Modification of the order of the display type/query parameters

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/complex-navigator-utils.ts:
```diff
@@ -0,0 +1,59 @@
+    if (minValue === maxValue) {
+      return minValue.toString();
+    } else {
+      return `${minValue}, ${maxValue}`;
```
```
return `${minValue}-${maxValue}`;
```

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-main-interactor/table-main-interactor.component.css:
```diff
@@ -0,0 +1,17 @@
+}
+
+.verticalLine {
+  border-right: 5px solid #0e6f76;
```
Change 5px to 3px to make the lines thiner.
```
  border-right: 3px solid #0e6f76;
```

jmedinaebi commented on src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-subcomponent-interactor/table-subcomponent-interactor.component.css:
```diff
@@ -0,0 +1,17 @@
+  z-index: 0;
+}
+
+.verticalLine {
```
Same as above, change 5px to 3px to make the lines thiner.
```
  border-right: 3px solid #0e6f76;
```

jmedinaebi commented on src/app/complex/complex-results/complex-results.component.ts:
```diff
@@ -194,4 +242,47 @@ export class ComplexResultsComponent implements OnInit, AfterViewInit {
+  }
+
+
+  setFirstDisplayType(): void {
```
To fix the redirect, what we need to do is to only redirect when people do a search. When people do a search, we don't have the display type set up, then we can redirect. When people change a filter, or change pages, or change from list to navigator or the other way around, display type is set, and we don't need to redirect.

So, that means, moving the redirect logic to inside the if `if (!this.DisplayType) {`

Basically, this method ends up like this
```
  setFirstDisplayType(): void {
    if (!this.DisplayType) {
      if (this._complexSearch.totalNumberOfResults === 1) {
        const complexId = this._complexSearch.elements[0].complexAC;
        if (!!complexId) {
          // For some reason this is needed so the navigate call works
          this.router.routeReuseStrategy.shouldReuseRoute = function () {
            return false;
          };
          this.router.navigate(['/complex', complexId]);
        }
      } else {
        // Currently the list view is the default, as we are just launching the navigator view
        // Later on we can change the default view to be the list or navigator view based on number of results
        if (this._complexSearch.totalNumberOfResults <= this._navigatorPageSize) {
          this.setComplexNavigatorView();
        } else {
          this.setListView();
        }
      }
    }
  }
```

jmedinaebi committed:
Fix bug with nested subcomplexes

jmedinaebi committed:
Fix stoichiometry formatting

jmedinaebi committed:
Make lines thinner

jmedinaebi committed:
Fix issues with filters and redirect

susiehgt committed:
Merge pull request #215 from Complex-Portal/complex-navigator-small-changes

Complex navigator small changes

susiehgt committed:
Changing name of variable into the header part to make it more logical;
Hidding of the orthology button;
Exchanging the magnifying glass to the external link button in compact mode

susiehgt committed:
fixing of the width of interactors column which was changing when subcomponents expanded

susiehgt committed:
fixing function name

jmedinaebi commented on src/app/complex/complex-results/complex-results.component.ts:
```diff
@@ -5,44 +5,61 @@ import {ComplexPortalService} from '../shared/service/complex-portal.service';
+  toast;
+
+  private _listPageSize = 15; // This is where we set the size of the pages for list view
+  private _navigatorPageSize = 15; // This is where we set the size of the pages for navigator view
```
Do we want to leave the paginator size in 15? Or something larger?

jmedinaebi commented on src/app/complex/complex-results/complex-results.component.ts:
```diff
@@ -153,22 +192,31 @@ export class ComplexResultsComponent implements OnInit, AfterViewInit {
 
   set complexSearch(value: ComplexSearchResult) {
     this._complexSearch = value;
+    this.setFirstDisplayType(); // if only one complex is in the search result, the display is list
```
This comment is outdated. If only one complex, then we redirect to the detailed complex page

susiehgt committed:
Change of the number of complexes displayed to 20
